### PR TITLE
Feature/logprobs seq2seq

### PR DESCRIPTION
**Pull request by jonathangomesselman**
***What existing issue does this pull request close?***

This addresses the following [shortcut ticket](https://app.shortcut.com/galileo/story/7119/dq-top-k-logprobs) and adds some more features.

Specifically this PR addresses:
- logging `gold_logprobs` rather than `gold_probs`
- Computing per sample DEP using `avg` rather than `max`
- Extracting and logging of the topk logprobs per token. 
- Computing sample perplexity

With these additions, I have re-architected the processing logic to better decompose the steps that can be shared between e.g. computing `gold_logprobs`, `top_logprobs` and `token_dep`.

***How are these changes tested?***

In the process of adding tests to `test_seq2seq.py`

### CODE REVIEW

File: dataquality/loggers/model_logger/seq2seq.py [hunk 1]
```diff
@@ -63,94 +67,241 @@ def validate_and_format(self) -> None:
...
+            batch_token_top_logprobs.append(token_top_logprobs)
+            batch_token_deps.append(batch_token_deps)
+
+            # Compute sample DEP as avg(token_DEP)
+            sample_dep = np.mean(batch_token_deps)
+            batch_deps.append(sample_dep)
+
+            # Perplexity = exp(-sum(gold_logprobs)
+            # TODO: Should this be in runners?
+            perplexity = np.exp(-1 * np.mean(token_gold_logprobs))
```
Comment (elboy3) [inline dataquality/loggers/model_logger/seq2seq.py:295]: if it's calculated on the full gold_logprobs and not just the top logprobs it should be here

but it should be an ML helper fn `def calculate_perplexity`

File: dataquality/loggers/model_logger/seq2seq.py [hunk 2]
```diff
@@ -63,94 +67,241 @@ def validate_and_format(self) -> None:
...
         """
+        # Compute the top-k logprob indices per token in the batch.
+        # We use np.argpartition to reduce the overhead of sorting the whole vocab.
+        # For reference see: https://numpy.org/doc/stable/reference/generated/numpy.argpartition.html
+        # Note: We multiply by -1 to reverse the order to help selecting the max
+        # TODO - this can be massively sped up using torch.top_k on the GPU
+        # TODO - use a parameter k not 5!
+        batch_logprobs *= -1
+        partitioned_logprob_indices = np.argpartition(batch_logprobs, 5, axis=-1)
+        top_logprob_indices = partitioned_logprob_indices[..., :5]
```
Comment (elboy3) [inline dataquality/loggers/model_logger/seq2seq.py:266]: nice!

File: dataquality/loggers/model_logger/seq2seq.py [hunk 3]
```diff
@@ -63,94 +67,241 @@ def validate_and_format(self) -> None:
...
 
-        First, extract the probabilities of the GT token label
+        return gold_logprobs, top_logprobs, token_deps
 
-        Probs is a numpy array of shape [batch_size, max_token_len, vocab_size] where
-        for each sample (text input) in the batch, every token of that sample has a
-        probability vector of size vocab_size (which can be 30k+).
+    def process_logprobs(
+        self, batch_ids: np.ndarray, batch_logprobs: np.ndarray
+    ) -> Tuple[pa.array, pa.array, pa.array, List[float], List[float]]:
```
Comment (elboy3) [inline dataquality/loggers/model_logger/seq2seq.py:211]: we should make this a type so it's not a massive tuple

File: dataquality/loggers/model_logger/seq2seq.py (hunk 1 above)
Comment (jonathangomesselman) [reply dataquality/loggers/model_logger/seq2seq.py]: It is computed using just the `gold_logprob` - hence I think it should not be here.

File: dataquality/utils/seq2seq.py [hunk 4]
```diff
@@ -1,11 +1,183 @@
...
+) -> Tuple[List[int], np.array, List[List[Tuple[str, int]]]]:
+    """Does a number of things that we lay out with comments"""
+    # 0) Get generated output ids
+    # TODO use convert_tensor_ndarray
+    gen_ids = generation_response.sequences[0].cpu().numpy()
+
+    # 1) Stack and convert to numpy
+    # When stacking squeeze the middle dim = [seq_len, vocab]
+    logits = torch.stack(generation_response.scores).squeeze()
+    logits = logits.cpu().numpy()
```
Comment (bogdan-galileo) [inline dataquality/utils/seq2seq.py:29]: detach too ?

File: dataquality/utils/seq2seq.py [hunk 5]
```diff
@@ -1,11 +1,183 @@
...
+    # 4) Remove eos token if present
+    if gen_ids[-1] == tokenizer.eos_token_id:
+        gen_ids = gen_ids[:-1]
+        logprobs = logprobs[:-1]
+
+    # 5) Compute token_probs (transition scores lol)
+    token_probs = np.take_along_axis(logprobs, gen_ids[..., None], axis=-1).squeeze()
+
+    # 6) Compute top_k ids
+    logprobs *= -1
```
Comment (bogdan-galileo) [inline dataquality/utils/seq2seq.py:47]: Note that this only returns the top 5 probs, but not sorted

**Comment by bogdan-galileo** [comment]
This branch can go 🗑️  ?

**Comment by jonathangomesselman** [comment]
Yeah I think. But maybe keep just as refference till we get everything fixed up

**Comment by elboy3** [comment]
Closing since we've merged this work in other PRs
